backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportHeaderGenerator.kt

```diff
@@ -1019,49 +1021,36 @@ private fun ObjCExportNamer.ClassOrProtocolName.toNameAttributes(): List<String>
 private fun swiftNameAttribute(swiftName: String) = "swift_name(\"$swiftName\")"
 private fun objcRuntimeNameAttribute(name: String) = "objc_runtime_name(\"$name\")"
 
-interface TypeParameterProvider{
-    fun typeAvailable(kotlinType: KotlinType):Boolean
+interface ObjCExportScope{
+    fun getName(typeParameterDescriptor: TypeParameterDescriptor?): String?
 }
 
-internal class ClassParameterProvider(container:DeclarationDescriptor): TypeParameterProvider {
-    constructor(method: FunctionDescriptor) : this(method.containingDeclaration)
-    private val nameSet:Set<String>
+internal class ObjCClassExportScope constructor(method: FunctionDescriptor): ObjCExportScope {
+    private val types:List<TypeParameterDescriptor>
     init {
-        nameSet = mutableSetOf()
-        if(container is ClassDescriptor && !container.isInterface) {
-            container.declaredTypeParameters.forEach {
-                nameSet.add(it.name.asString())
-            }
+        val container = method.containingDeclaration
+        types = if(container is ClassDescriptor && !container.isInterface) {
+            container.declaredTypeParameters
+        } else {
+            emptyList()
         }
     }
 
-    override fun typeAvailable(kotlinType: KotlinType): Boolean =
-            nameSet.contains(kotlinType.typeParameterName())
+    override fun getName(typeParameterDescriptor: TypeParameterDescriptor?): String? =
+            if(typeParameterDescriptor != null && types.contains(typeParameterDescriptor)){
+                typeParameterDescriptor.name.asString()
```

Comment:
Yeah, interesting. Xcode seems to be OK with `id`, which is a surprise, but it should be renamed. I want to think about that a bit to see if there are other problematic names.

Reply:
> Xcode seems to be OK with id, which is a surprise

`id` is just a type name. It is not forbidden to reuse such names for anything else.
The problem here is that we are trying to refer `id` in `foo`, i.e. from the scope where it is overridden by type parameter.

> if there are other problematic names.

Sure!
Potentially type parameter name can override name of any type. E.g.
0) Some "built-in" types emitted (like `id`)
1) Objective-C types imported to Kotlin (e.g. `NSObject`) when exported to Objective-C header
2) Kotlin class names (not that serious, because Kotlin class names are prefixed)

Reply:
Went through a couple iterations on 2. I was hoping to delay writing names until the end as it looks like stubs delay rendering, but they can evaluate earlier, so that didn't work. Then I landed on the far easier method of checking type param names and objc class/protocol names, and appending underscores to either. So, in the following:

```kotlin
class GenClashNames<ValuesGenericsClashnameClass, ValuesGenericsClashnameProtocol, ValuesGenericsValues_genericsKt>() {
    fun foo(): Any = ClashnameClass("nnn")
}

data class ClashnameClass(val str: String)
interface ClashnameProtocol {
    val str: String
}
```

```objc
@interface ValuesGenericsGenClashNames<ValuesGenericsClashnameClass, 
```

The type param name is above, the class name written later is:

```objc
@interface ValuesGenericsClashnameClass_ : KotlinBase
```
This is rare, obviously, and swift names don't change. Let me know what you think. The alternatives seemed like a lot of work for little gain.

Reply:
This is ok for now.

Reply:
After working with that for a while, it turns out that if you have a type parameter name that clashes with a class name, using the target class in the generic means the target class's name will eval before the generic, and nothing will clash. I've been trying to cause a test issue but have been unable to, so I rolled back some of what I said above.

The type parameter name will be mangled, but not the class name itself.

Added some tests to hopefully catch a failing issue if that code changes in the future.

```kotlin
class GenClashNames<ValuesGenericsClashnameClass, ValuesGenericsClashnameProtocol, ValuesGenericsClashnameParam, ValuesGenericsValues_genericsKt>() {
    fun foo() = ClashnameClass("nnn")

    fun bar(): ClashnameProtocol = object : ClashnameProtocol{
        override val str = "qqq"
    }

    fun baz(arg: ClashnameParam): Boolean {
        return arg.str == "meh"
    }
}

open class ClashnameClass(val str: String)
interface ClashnameProtocol {
    val str: String
}
data class ClashnameParam(val str: String)
```

produces

```objc
@interface ValuesGenericsGenClashNames<ValuesGenericsClashnameClass_, ValuesGenericsClashnameProtocol_, ValuesGenericsClashnameParam_, ValuesGenericsValues_genericsKt> : KotlinBase
- (instancetype)init __attribute__((swift_name("init()"))) __attribute__((objc_designated_initializer));
+ (instancetype)new __attribute__((availability(swift, unavailable, message="use object initializers instead")));
- (ValuesGenericsClashnameClass *)foo __attribute__((swift_name("foo()")));
- (id<ValuesGenericsClashnameProtocol>)bar __attribute__((swift_name("bar()")));
- (BOOL)bazArg:(ValuesGenericsClashnameParam *)arg __attribute__((swift_name("baz(arg:)")));
@end;

__attribute__((swift_name("ClashnameClass")))
@interface ValuesGenericsClashnameClass : KotlinBase
- (instancetype)initWithStr:(NSString *)str __attribute__((swift_name("init(str:)"))) __attribute__((objc_designated_initializer));
@property (readonly) NSString *str __attribute__((swift_name("str")));
@end;

__attribute__((swift_name("ClashnameProtocol")))
@protocol ValuesGenericsClashnameProtocol
@required
@property (readonly) NSString *str __attribute__((swift_name("str")));
@end;

__attribute__((objc_subclassing_restricted))
__attribute__((swift_name("ClashnameParam")))
@interface ValuesGenericsClashnameParam : KotlinBase
- (instancetype)initWithStr:(NSString *)str __attribute__((swift_name("init(str:)"))) __attribute__((objc_designated_initializer));
- (BOOL)isEqual:(id _Nullable)other __attribute__((swift_name("isEqual(:)")));
- (NSUInteger)hash __attribute__((swift_name("hash()")));
- (NSString *)description __attribute__((swift_name("description()")));
- (NSString *)component1 __attribute__((swift_name("component1()")));
- (ValuesGenericsClashnameParam *)doCopyStr:(NSString *)str __attribute__((swift_name("doCopy(str:)")));
@property (readonly) NSString *str __attribute__((swift_name("str")));
@end;

Reply:
How have you implemented this type parameter name mangling logic? Which part of code is it located at, and which data does it use?

> using the target class in the generic means the target class's name will eval before the generic

Could you clarify?

Reply:
Still need to push. Wanted a more complete implementation before pushing and have been distracted with other concurrent work. Will push now, but am still poking around some.

I have implemented this. It's in `ObjCExportNamerImpl`. It does not extend from `Mapping`, as the concept of `nameToElements` didn't make a lot of sense, but that means there's some code duplication. To refactor possibly.

To clarify the statement, if the type param name is `ValuesGenericsClashnameClass`, that will clash with `ClashnameClass`. That means we want to rename the param to `ValuesGenericsClashnameClass_`, but that's only a problem if we're using that target class, `ClashnameClass`, in the generic parent class (`GenClashNames` in the example above). If we encountered the type param name first, then the class itself would need a name mangling, but because the generator code evaluates properties, methods, and parent classes first, we always resolve the target class's name before we need to assign that type param, so the target class's name itself never needs mangling because of the type param name clashing.

Summary, the type param may be mangled because is clashes with a class name, but not the other way around.

Reply:
Also, to clarify, "Wanted a more complete implementation". I'm just cleaning up tests and had planned to push last night, but wanted better test coverage and also want a full review that with the experimental flag off that nothing is changed, but the implementation is "complete" assuming I don't find anything with tests and review.

Reply:
Ok, thank you.

> the generator code evaluates properties, methods, and parent classes first

Yes, this is the detail I missed.

> Summary, the type param may be mangled because is clashes with a class name, but not the other way around.

That's great!

Reply:
> the generator code evaluates properties, methods, and parent classes first

Btw, what if type parameter is used in properties or methods?

```kotlin
class Foo<ValuesGenericsBar> {
    fun foo(p: ValuesGenericsBar) {}
    fun bar(p: Bar) {}
}

class Bar
```

->

```obj-c
__attribute__((objc_subclassing_restricted))
__attribute__((swift_name("Foo")))
@interface ValuesGenericsFoo<ValuesGenericsBar> : KotlinBase
// ...
- (void)fooP:(ValuesGenericsBar _Nullable)p __attribute__((swift_name("foo(p:)")));
- (void)barP:(ValuesGenericsBar *)p __attribute__((swift_name("bar(p:)")));
@end;
```

I'm still not sure if it makes much sense to fix this now though.

Reply:
Well, I'm not happy that I missed that. My testing got caught up in the outer class definition. Anyway, the RenderedStub will render the name as `buildAsDeclaredOrInheritedMethods` creates a Set, which hash/equals results in rendering the underlying stub. To avoid this, we'd either need run a pass through the classes properties and methods before calling `collectMethodsOrProperties` (or some similar first pass, perhaps higher up), or mangle the class name.

It is obviously unlikely that names would clash, and if we're expecting significant refactoring before generics would be standard, if ever, then fixing later is OK, as long as it's addressed at some point.

Let me know if you want me to deal with it now. I'd be inclined to pass through class members in `StubBuilder.translateClassMembers` to flush out name issues.

Reply:
I find it ok to stop at this point.
We have some plans to fully rework clash handling in generated header, and I believe that we'll handle then these clashes between type parameters and classes.